auth/standard_role_manager.cc

```diff
@@ -250,26 +251,26 @@ future<> standard_role_manager::start() {
             }
             return make_ready_future<>();
         }).then([this] {
-            _stopped = auth::do_after_system_ready(_as, [this] {
-                return seastar::async([this] {
-                    if (legacy_mode(_qp)) {
-                        _migration_manager.wait_for_schema_agreement(_qp.db().real_database(), db::timeout_clock::time_point::max(), &_as).get();
-
-                        if (any_nondefault_role_row_satisfies(_qp, &has_can_login).get()) {
-                            if (legacy_metadata_exists()) {
-                                log.warn("Ignoring legacy user metadata since nondefault roles already exist.");
-                            }
-
-                            return;
-                        }
+            _stopped = auth::do_after_system_ready(_as, [this] () -> future<>  {
+                auto signal_super_user_created = defer([this] {
```

Comment:
is this for sure correct scope? maybe there is no good place to place defer, keep in mind that do_after_system_ready is potentially calling lambda several times due to retries

there are probably multiple ways to solve this, for instance attach it to the future returned by do_system_ready, like:

```
auth::do_after_system_ready(...).then_wrapped(future<> f) {
try {
f.get()
catch(...) { ...broken(); return; }
...signal()
});
```

Reply:
yes, no need to defer because if try block you will anyway have it in one place (after try block) and that was the point of using defer

Reply:
also you should be able to pass exception_ptr to broken(), so that it will be acted on startup instead as currently only on ::stop()

But *do* check how exception lifetime works as I am not sure if exception_ptr is smart pointer...

Reply:
yes I didn't mean you should not throw but do both